lib/rate.c

```diff
@@ -0,0 +1,154 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "picotls.h"
+#include "quicly/rate.h"
+
+static void start_sampling(quicly_ratemeter_t *meter, int64_t now, uint64_t bytes_acked)
+{
+    meter->current.start.at = now;
+    meter->current.start.bytes_acked = bytes_acked;
+}
+
+static void commit_sample(quicly_ratemeter_t *meter)
+{
+    ++meter->past_samples.latest;
+    if (meter->past_samples.latest >= PTLS_ELEMENTSOF(meter->past_samples.entries))
+        meter->past_samples.latest = 0;
+    meter->past_samples.entries[meter->past_samples.latest] = meter->current.sample;
+
+    meter->current.start.at = INT64_MAX;
+    meter->current.sample = (struct st_quicly_rate_sample_t){};
+}
+
+void quicly_ratemeter_init(quicly_ratemeter_t *meter)
+{
+    *meter = (quicly_ratemeter_t){
+        .past_samples = {.latest = PTLS_ELEMENTSOF(meter->past_samples.entries) - 1},
+        .pn_cwnd_limited = {.start = UINT64_MAX, .end = UINT64_MAX},
+        .current = {.start = {.at = INT64_MAX}},
+    };
+}
+
+void quicly_ratemeter_in_cwnd_limited(quicly_ratemeter_t *meter, uint64_t pn)
+{
+    /* bail out if already in cwnd-limited phase */
+    if (meter->pn_cwnd_limited.start != UINT64_MAX && meter->pn_cwnd_limited.end == UINT64_MAX)
+        return;
+
+    /* if the estimator was waiting for the end of the previous phase, and if a valid partial sample exists, commit it now */
+    if (meter->pn_cwnd_limited.end != UINT64_MAX && meter->current.sample.elapsed != 0)
+        commit_sample(meter);
+
+    /* begin new cwnd-limited phase */
+    meter->pn_cwnd_limited = (quicly_range_t){.start = pn, .end = UINT64_MAX};
+}
+
+void quicly_ratemeter_not_cwnd_limited(quicly_ratemeter_t *meter, uint64_t pn)
+{
+    if (meter->pn_cwnd_limited.start != UINT64_MAX && meter->pn_cwnd_limited.end == UINT64_MAX)
+        meter->pn_cwnd_limited.end = pn;
+}
+
+void quicly_ratemeter_on_ack(quicly_ratemeter_t *meter, int64_t now, uint64_t bytes_acked, uint64_t pn)
+{
+    if (meter->pn_cwnd_limited.start <= pn && pn < meter->pn_cwnd_limited.end) {
+        /* At the moment, the flow is CWND-limited. Either start the timer or update. */
+        if (meter->current.start.at == INT64_MAX) {
+            start_sampling(meter, now, bytes_acked);
+        } else {
```

Comment:
Add comment here telling that we update the sample so that it can be committed other than when receiving an ACK.

Reply:
Fixed in 44fa694.